s3tests_boto3/functional/test_iam.py

```diff
@@ -0,0 +1,939 @@
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    e = assert_raises(ClientError, client.delete_user_policy, PolicyName='non-existing-policy-name',
+                      UserName=get_tenant_user_id())
+    status = _get_status(e.response)
+    eq(status, 404)
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+
+@attr(resource='user-policy')
+@attr(method='delete')
+@attr(operation='Verify Delete multiple User policies for a user')
+@attr(assertion='succeeds')
+@attr('user-policy')
+def test_delete_user_policy_from_multiple_policies():
+    client = get_iam_client()
+
+    policy_document_allow = json.dumps(
+        {"Version": "2012-10-17",
+         "Statement": {
+             "Effect": "Allow",
+             "Action": "*",
+             "Resource": "*"}}
+    )
+
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowAccessPolicy1',
+                                      UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowAccessPolicy2',
+                                      UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.put_user_policy(PolicyDocument=policy_document_allow,
+                                      PolicyName='AllowAccessPolicy3',
+                                      UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy1',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy2',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+    response = client.get_user_policy(PolicyName='AllowAccessPolicy3',
+                                      UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+    response = client.delete_user_policy(PolicyName='AllowAccessPolicy3',
+                                         UserName=get_tenant_user_id())
+    eq(response['ResponseMetadata']['HTTPStatusCode'], 200)
+
+
+@attr(resource='user-policy')
+@attr(method='s3 Actions')
+@attr(operation='Verify Allow Bucket Actions in user Policy')
+@attr(assertion='succeeds')
+@attr('user-policy')
+def test_allow_bucket_actions_in_user_policy():
+    client = get_iam_client()
+    s3_client_tenant = get_tenant_client()
+
+    s3_client_iam = get_iam_s3client()
```

Comment:
Here again the bucket is created by the iam_user (via s3 api), and the policy is attached to the iam_user that is creating the bucket right? and then the bucket is being accessed by a user in a different tenant. Has this been tested against RGW?

Reply:
no, sorry the policy is attached to the tenant user. And now the tenant user tries to access the bucket which is in the tenant of the iam user and these two tenants are different right? Does this pass against RGW?

Reply:
> 1. policy

You may have to fix these discrepancies in RGW.

Reply:
> Yes, this passes against RGW.
> 
> But we are seeing some issues with RGW
> 
> 1. when we try to get non existing policy (or deleted policy) it return malformed response which boto3 is unable to decode where as AWS returns "NoSuchEntity"
> 2. Policy without statement gets accepted where as AWS returns error "MalformedPolicyDocument"
> 3. Policy with two same Statement IDs in a single policy gets accepted where as AWS returns " MalformedPolicyDocument"

For all three, you will have to fix the code.
> 
> Due to above three issues, three tests are failing on RGW
> 
> 1. test_put_user_policy_invalid_element
> 2. test_get_user_policy_invalid_policy_name
> 3. test_get_deleted_user_policy

Reply:
> Are you sure that this passes? because it doesnt for me on the latest master. I have the exact same test where I create an iam_client in 'empty' tenant and attach a policy to a user in a tenant 'testx'. Then I create a bucket and upload an object accessing the 's3' api using the iam_client. Then I create an s3 connection using the user in tenant 'testx', and try to list objects - ListBucket fails with error - Bucket not found and rightly so, since when the bucket you have created using iam_client is under 'empty' tenant, and when the tenant user tries to access it, the bucket is searched in 'testx' tenant and this fails. The following are the logs in rgw, for list_bucket: 2022-06-03T20:46:27.660+0530 7f3a61024700 10 req 3483647770686675559 0.001000012s s->object= s->bucket=testx/my-new-shiny-bucket 2022-06-03T20:46:27.660+0530 7f3a61024700 2 req 3483647770686675559 0.001000012s s3:list_bucket init permissions 2022-06-03T20:46:27.660+0530 7f3a61024700 20 req 3483647770686675559 0.001000012s **s3:list_bucket get_system_obj_state: rctx=0x7f39ec6372b0 obj=default.rgw.meta:root:testx/my-new-shiny-bucket state=0x55801ad1e0a0 s->prefetch_data=0 2022-06-03T20:46:27.660+0530 7f3a61024700 10 req 3483647770686675559 0.001000012s** s3:list_bucket cache get: name=default.rgw.meta+root+testx/my-new-shiny-bucket : miss ... **2022-06-03T20:46:27.661+0530 7f3a5d81d700 5 bucket my-new-shiny-bucket doesn't exist, not logging**
> 
> Can you verify that the tenants of the iam_client and the tenanted_user are the same or not, if they are then this is a valid test and it must pass, else it must not.

Just for your reference, I am attaching the rgw logs for create_bucket also here:
2022-06-03T20:42:26.829+0530 7f3aaf8c1700  2 req 10478071630861750662 0.003000033s s3:create_bucket normalizing buckets and tenants
**2022-06-03T20:42:26.829+0530 7f3aaf8c1700 10 req 10478071630861750662 0.003000033s s->object=<NULL> s->bucket=my-new-shiny-bucket**
2022-06-03T20:42:26.831+0530 7f3aae0be700 10 req 10478071630861750662 0.005000055s s3:create_bucket cache put: name=default.rgw.meta+root+my-new-shiny-bucket info.flags=0x0
**2022-06-03T20:42:26.831+0530 7f3aae0be700 10 req 10478071630861750662 0.005000055s s3:create_bucket adding default.rgw.meta+root+my-new-shiny-bucket to cache LRU end**
2022-06-03T20:42:26.831+0530 7f3aae0be700 10 req 10478071630861750662 0.005000055s s3:create_bucket user=TESTER bucket=:my-new-shiny-bucket[])

You can see that in create_bucket there is no tenant: default.rgw.meta+root+my-new-shiny-bucket
Whereas, in list_bucket it searches for the bucket under tenant 'testx': default.rgw.meta:root:testx/my-new-shiny-bucket
Hence the search fails. This just for you to verify what the test is actually doing.

Reply:
Yes, both the users were in same tenant and now using s3 alt and iam user which are supposed to be under same tenant.
Also fixed the code for failing tests.